projects/baby_driver/test/test_spi_exchange.c

```diff
@@ -0,0 +1,354 @@
+  test_equal_multiple_messages();
+
+  TEST_ASSERT_EQUAL(5, s_callback_counter_status);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data_status[0]);
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, s_received_data_status[1]);
+
+  clear_received_data_array();
+  s_num_messages = 0;
+}
+
+void test_invalid_input(void) {
+  // first test invalid cs port
+  send_first_meta_data(0, 0, 7, 7, INVALID_CS_PORT, 0, 0);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(1, s_callback_counter_status);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data_status[0]);
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, s_received_data_status[1]);
+
+  // second test invalid cs pin
+  send_first_meta_data(0, 0, 7, 7, 0, INVALID_CS_PIN, 0);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(2, s_callback_counter_status);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data_status[0]);
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, s_received_data_status[1]);
+
+  // third test invalid port
+  send_first_meta_data(INVALID_SPI_PORT, 0, 7, 7, 0, 0, 0);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(3, s_callback_counter_status);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data_status[0]);
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, s_received_data_status[1]);
+
+  // fourth test invalid mode
+  send_first_meta_data(0, INVALID_SPI_MODE, 7, 7, 0, 0, 0);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(4, s_callback_counter_status);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data_status[0]);
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, s_received_data_status[1]);
+}
+
+void test_timeout(void) {
+  // test for watchdog error with second meta data message
+  send_first_meta_data(0, 0, 7, 7, 0, 0, 0);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(1, s_callback_counter_status);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data_status[0]);
+  TEST_ASSERT_EQUAL(STATUS_CODE_TIMEOUT, s_received_data_status[1]);
+
+  // test for watchdog error with data message
+  send_meta_data(0, 0, 7, 7, 0, 0, 0, 0x005b8d80);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(2, s_callback_counter_status);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data_status[0]);
+  TEST_ASSERT_EQUAL(STATUS_CODE_TIMEOUT, s_received_data_status[1]);
+}
+
+StatusCode TEST_MOCK(spi_exchange)(SpiPort spi, uint8_t *tx_data, size_t tx_len, uint8_t *rx_data,
+                                   size_t rx_len) {
+  for (uint8_t i = 0; i < rx_len; i++) rx_data[i] = 1;
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode TEST_MOCK(spi_init)(SpiPort spi, const SpiSettings *settings) {
+  // still need to add tests making sure the settings and port are the same
```

Comment:
Do you still?

Reply:
Oh I thought I deleted this comment, I finished that.